### PR TITLE
command: Add `terraform untaint`

### LINKED ISSUE #4820: core: Add support for "terraform untaint" command (undo taint)
When user taints a resource, it is very difficult (from user's point of view basically impossible) to undo that step.

Therefore I think that having either `terraform untaint` or `terraform taint -undo` (final syntax is not really the concern) to undo what user has done via `terraform taint` would be useful.

### PR DESCRIPTION
**Pull request by phinze**
- [x] Docs
 - [x] Command Unit Tests
 - [x] State Unit Tests

Closes #4820

### CODE REVIEW

**Comment by phinze** [comment]
Want to fill in a few more unit tests for `(r *ResourceState) Untaint(index) error` but this is ready for initial review at least!

File: command/untaint.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	Meta
+}
+
+func (c *UntaintCommand) Run(args []string) int {
+	args = c.Meta.process(args, false)
+
+	var allowMissing bool
+	var module string
+	var index int
+	cmdFlags := c.Meta.flagSet("taint")
```
Comment (radeksimko) [inline command/untaint.go]: Shouldn't the flag set name be `untaint`? I guess it's just typo.

File: command/untaint.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	cmdFlags.StringVar(&c.Meta.statePath, "state", DefaultStateFilename, "path")
+	cmdFlags.StringVar(&c.Meta.stateOutPath, "state-out", "", "path")
+	cmdFlags.StringVar(&c.Meta.backupPath, "backup", "", "path")
+	cmdFlags.IntVar(&index, "index", -1, "index")
+	cmdFlags.Usage = func() { c.Ui.Error(c.Help()) }
+	if err := cmdFlags.Parse(args); err != nil {
+		return 1
+	}
+
+	// Require the one argument for the resource to taint
```
Comment (radeksimko) [inline command/untaint.go]: > _... for the resource to taint_

for the resource to **untaint**

File: command/untaint.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+		c.Ui.Error(fmt.Sprintf(
+			"The state is empty. The most common reason for this is that\n" +
+				"an invalid state file path was given or Terraform has never\n " +
+				"been run for this infrastructure. Infrastructure must exist\n" +
+				"for it to be untainted."))
+		return 1
+	}
+
+	// Get the proper module we want to taint
```
Comment (radeksimko) [inline command/untaint.go]: > _module we want to taint_

module we want to **untaint**

File: command/untaint.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	modPath := strings.Split(module, ".")
+	mod := s.ModuleByPath(modPath)
+	if mod == nil {
+		if allowMissing {
+			return c.allowMissingExit(name, module)
+		}
+
+		c.Ui.Error(fmt.Sprintf(
+			"The module %s could not be found. There is nothing to taint.",
```
Comment (radeksimko) [inline command/untaint.go]: > There is nothing to taint.

There is nothing to **untaint.**

File: command/untaint.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	// If there are no resources in this module, it is an error
+	if len(mod.Resources) == 0 {
+		if allowMissing {
+			return c.allowMissingExit(name, module)
+		}
+
+		c.Ui.Error(fmt.Sprintf(
+			"The module %s has no resources. There is nothing to taint.",
```
Comment (radeksimko) [inline command/untaint.go]: same here

File: command/untaint.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	helpText := `
+Usage: terraform untaint [options] name
+
+  Manually unmark a resource as tainted, restoring it as the primary
+	instance in the state.  This reverses either a manual 'terraform taint' or
+	the result of provisioners failing on a resource.
+
+  This will not modify your infrastructure. This command changes your
+  state to unmark a resource as tainted so that during the next plan or
+  apply, that resource will be destroyed and recreated. This command on
```
Comment (radeksimko) [inline command/untaint.go]: > during the next plan or
>  apply, that resource will be destroyed and recreated

I guess that's a copy-paste mistake - untaint will do the exact opposite.

**Comment by phinze** [comment]
Thanks for the review, @radeksimko - addressed the typos you caught and added the state unit tests I mentioned. PTAL when you have a moment!

**Comment by phinze** [comment]
@radeksimko ping :grinning:

**Comment by radeksimko** [comment]
@phinze just looking at this...

I first had to look up `PTAL` abbreviation... I first found `Post-Traumatic Activity Level (head injury)`, so I searched further to find `Please Take Another Look`. :smile:

File: command/untaint_test.go [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+import (
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/mitchellh/cli"
+)
+
+func TestUntaint(t *testing.T) {
```
Comment (radeksimko) [inline command/untaint_test.go]: Just a nitpick, but as this is a test testing noop case, can we also name it that way, so it's obvious?

File: command/untaint.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+import (
+	"fmt"
+	"log"
+	"strings"
+)
+
+// UntaintCommand is a cli.Command implementation that manually untaints
+// a resource, marking it as primary and ready for service.
+type UntaintCommand struct {
```
Comment (radeksimko) [inline command/untaint.go]: We'll also need to add this into `commands.go`, otherwise `terraform untaint` won't work.

File: terraform/state.go [hunk 9]
```diff
@@ -872,6 +872,36 @@ func (r *ResourceState) Taint() {
 	r.Primary = nil
 }
 
+// Untaint takes a tainted InstanceState and marks it as primary.
+// The index argument is used to select a single InstanceState from the
+// array of Tainted when there are more than one. If index is -1, the
+// first Tainted InstanceState will be untainted iff there is only one
```
Comment (radeksimko) [inline terraform/state.go]: typo: `iff` -> `if`

File: terraform/state.go [hunk 10]
```diff
@@ -872,6 +872,36 @@ func (r *ResourceState) Taint() {
...
+	if len(r.Tainted) == 0 {
+		return fmt.Errorf("Nothing to untaint.")
+	}
+	if r.Primary != nil {
+		return fmt.Errorf("Resource has a primary instance in the state that would be overwritten by untainting. If you want to restore a tainted resource to primary, taint the existing primary instance first.")
+	}
+	if index == -1 && len(r.Tainted) > 1 {
+		return fmt.Errorf("There are %d tainted instances for this resource, please specify an index to select which one to untaint.", len(r.Tainted))
+	}
+	if index == -1 {
```
Comment (radeksimko) [inline terraform/state.go]: Is there any specific reason why we want to support `-1`?

File: terraform/state.go (hunk 9 above)
Comment (phinze) [inline terraform/state.go]: I was using it in the `iff = "if and only if"` meaning.

File: terraform/state.go (hunk 10 above)
Comment (phinze) [inline terraform/state.go]: This is just an internal mechanism to get the behavior:

 * Index can be omitted when there is only one tainted resource
 * Index is required when there is more than one tainted resource

I'd welcome ideas for other ways to achieve those properties. :+1:

File: command/untaint.go (hunk 8 above)
Comment (phinze) [inline command/untaint.go]: Oh ha! Fixing now.

File: terraform/state.go (hunk 9 above)
Comment (radeksimko) [inline terraform/state.go]: ah, right! :+1:

File: terraform/state.go (hunk 10 above)
Comment (radeksimko) [inline terraform/state.go]: Ah, right, that's because `nil` "version" of integer in Go is zero. :smiley: This makes sense then.

**Comment by radeksimko** [comment]
Other than the test naming nitpick and `commands.go`, this LGTM. I compiled from your branch and tested this on real code, worked :ok_hand:

File: command/untaint_test.go (hunk 7 above)
Comment (phinze) [inline command/untaint_test.go]: I don't think this is a noop - it's the "basic" use case - `test_instance.foo` starts tainted and ends up untainted.

Comment (radeksimko) [inline command/untaint_test.go]: d'oh, obviously, sorry!

**Comment by radeksimko** [comment]
This LGTM now.

Feel free to merge.

**Pull request merged**